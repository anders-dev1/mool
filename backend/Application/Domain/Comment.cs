using System;
using System.Collections.Generic;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace Application.Domain
{
    public record Comment
    {
        [BsonId]
        public ObjectId Id { get; set; }
        public DateTimeOffset Created { get; set; }
        public string Content { get; set; }
        public ObjectId UserId { get; set; }
        public HashSet<ObjectId> LikedBy { get; set; }

        public Comment(
            string content,
            ObjectId userId,
            DateTimeOffset created)
        {
            Id = ObjectId.GenerateNewId();
            Created = created;
            Content = content;
            UserId = userId;
            LikedBy = new HashSet<ObjectId>();
        }
    }
}