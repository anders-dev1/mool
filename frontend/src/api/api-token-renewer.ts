import {AccessTokenClient, RenewAccessTokenCommand } from "./api-client";
import {store} from "../redux/store";
import {sessionActions} from "../redux/actions/session.action";

export async function RenewToken() {
    let refreshToken = localStorage.getItem("refresh-token");
    if (!refreshToken){
        return Promise.reject('There was no refresh token in storage.');
    }

    // use refresh token to get new access token
    let client = new AccessTokenClient();
    let request = new RenewAccessTokenCommand();
    request.refreshToken = refreshToken;

    await client.renew(request).then(
        result => {
            localStorage.setItem("access-token", result.accessToken);
        },
        error => {
            if (error.status === 404){
                store.dispatch(sessionActions.renewTokenExpired());
            }
        }
    );
}