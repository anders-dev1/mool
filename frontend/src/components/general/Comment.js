import React from "react";
import ProfilePictureComponent from "./ProfilePictureComponent";
import ActionPill from "./ActionPillComponent";
import {withTheme} from "styled-components";
import {connect} from "react-redux";
import {commentActions} from "../../redux/actions/comment.actions";

class Comment extends React.PureComponent {
    render() {
        return(
        <div>
            <div className={"flex justify-start"}>
                <ProfilePictureComponent className={"mr1"} isComment/>
                <div className={"flex-column pt1"}>
                    <p className={"sub"}>{this.props.comment.authorName}</p>
                    <p className={"sub mt1"} data-testid={"commentContent"}>{this.props.comment.content}</p>
                </div>
            </div>
            <div className={"flex justify-end"}>
                <ActionPill
                    icon={"👍"}
                    onClick={() => this.props.toggleCommentLike(this.props.commentId)}
                    number={this.props.comment.likes}
                    color={(this.props.comment.likedByCurrentUser
                        ? this.props.theme.successColor
                        : this.props.theme.primaryColor )}
                    className={"mr1"}
                    processing={this.props.comment.isChangingLikeStatus}
                    data-testid={"commentLikeBtn"}/>
            </div>
        </div>)
    }
}

function mapState(state, props) {
    const comment = state.commentsReducer.commentsById[props.commentId];
    return { comment }
}

const actionCreators = {
    toggleCommentLike: commentActions.toggleCommentLike
}

const connectedComponent = withTheme(connect(mapState, actionCreators)(Comment))
export {connectedComponent as Comment}
export * from './Comment';