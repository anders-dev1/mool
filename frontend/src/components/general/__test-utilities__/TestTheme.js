export const testTheme = {
    textColor: "black",
    primaryColor: "#007BFF",
    primaryHover: "rgba(0,123,255,0.33)",
    successColor: "#28A745",
    errorColor: "#F02849",
    backgroundColor: "#F0F2F5",
    buttonTextColor: "#FFFFFF",
    dividerColor: "#C7CBD1",
    inputColor: "#99A5B0",
    spinnerColor: "#FFFFFF transparent transparent transparent",
    font: "Acumin Pro, Arial, sans-serif !important;"
};