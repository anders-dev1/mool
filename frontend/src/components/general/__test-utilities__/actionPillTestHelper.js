import {screen} from "@testing-library/react";

export function findActionPillByIcon(icon) {
    return screen.getByText(icon).closest('div');
}

export function findActionPillNumberByIcon(icon) {
    const pill = findActionPillByIcon(icon);
    const numberP = pill.lastChild.lastChild;
    return numberP.textContent;
}