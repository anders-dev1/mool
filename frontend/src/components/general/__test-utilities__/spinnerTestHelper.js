export function findSpinner(container){
    return container.getElementsByClassName('lds-ring')[0];
}