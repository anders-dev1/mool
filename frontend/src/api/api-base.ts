import fetchIntercept, {FetchInterceptorResponse} from 'fetch-intercept';
import {RenewToken} from "./api-token-renewer";

class Request {
    url: string;
    config: Config;
    constructor(url: string, config: Config) {
        this.url = url;
        this.config = config;
    }
}

const ignoredPaths: string[] = [
    "/api/accesstoken/renew",
    "/api/accesstoken/authenticate"
]

export class ApiBase {
    constructor() {
        let originalRequest: Request;
        const interceptor = fetchIntercept.register({
            request(url: string, config: Config): Promise<any[]> | any[] {
                if (ignoredPaths.some((path) => url.endsWith(path))){
                    // Calling this function removes this interceptor from the list of interceptors.
                    // This stops our interceptor from processing any longer in the current request chain.
                    interceptor();
                    return [url, config];
                }

                originalRequest = new Request(url, config);

                const token = localStorage.getItem('access-token');
                if (!token){
                    return [url, config];
                }

                const myHeaders: HeadersInit = new Headers();
                myHeaders.set('Accept', 'application/json');
                myHeaders.set('Content-Type', 'application/json');
                myHeaders.set('Authorization', "Bearer " + token);

                config.headers = myHeaders;

                return [url, config];
            },
            response(response: FetchInterceptorResponse): Promise<FetchInterceptorResponse> | any {
                if (response.status != 401) {
                    return response;
                }

                if (originalRequest.url.includes('token')) {
                    return response;
                }

                let refreshToken = localStorage.getItem("refresh-token");
                if (!refreshToken) {
                    return response;
                }

                // use refresh token to get new access token
                return RenewToken().then(
                    result => {
                        const accessToken = localStorage.getItem('access-token');

                        const myHeaders: HeadersInit = new Headers();
                        myHeaders.set('Authorization', "Bearer " + accessToken);
                        myHeaders.set('Accept', 'application/json');
                        myHeaders.set('Content-Type', 'application/json');
                        originalRequest.config.headers = myHeaders;

                        return fetch(originalRequest.url, originalRequest.config).then(
                            result => {
                                return result;
                            }
                        );
                    }
                );
            }
        });
    }
}

export interface Config {
    method: string;
    headers: HeadersInit;
}