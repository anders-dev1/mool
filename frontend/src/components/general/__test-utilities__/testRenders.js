import React from 'react'
import {render} from '@testing-library/react'
import {Provider} from 'react-redux'
import {setupStore, store} from "../../../redux/store";
import {testTheme} from "./TestTheme";
import {ThemeProvider} from "styled-components";

// This is used to set up the store for tests.
export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        testStore = setupStore(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({children}) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return {testStore, ...render(ui, {wrapper: Wrapper, ...renderOptions})}
}

// This is used with the react-test-renderer to render dom in tests.
export function internalsForRenderer(
    ui,
    {
        preloadedState = {},
        testStore = setupStore(preloadedState)
    } = {}
){
    function Wrapper(children) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return Wrapper(ui);
}