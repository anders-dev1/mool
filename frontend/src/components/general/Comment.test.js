import renderer from "react-test-renderer";
import {Comment} from "./Comment";
import {testTheme} from "./__test-utilities__/TestTheme";
import {LikeCommentClient} from "../../api";
import {fireEvent, render, screen} from "@testing-library/react";
import {internalsForRenderer, renderWithProviders} from "./__test-utilities__/testRenders";

const likedComment = {
    id: 1,
    authorName: "Person Personsen",
    content: "This is the comment content.",
    likes: 42,
    likedByCurrentUser: true,
    isChangingLikeStatus: false
}

const unlikedComment = {
    id: 1,
    authorName: "This is the comment content.",
    content: "Lorem ipsum",
    likes: 42,
    likedByCurrentUser: false,
    isChangingLikeStatus: false
}

function createState(comment){
    return {
        commentsReducer: {
            commentsById:{
                1: comment
            }
        }
    }
}

beforeEach(() => {
    jest.spyOn(LikeCommentClient.prototype, 'unlike')
        .mockImplementation(() => {
            return Promise.resolve();
        });
    jest.spyOn(LikeCommentClient.prototype, 'like')
        .mockImplementation(() => {
            return Promise.resolve();
        });
});

function renderComment(comment) {
    renderWithProviders(
        <Comment commentId={1}/>,
        {preloadedState: createState(comment)});

    return screen.getByText("👍").parentElement;
}

test('renders correctly', () => {
    const internalDom = internalsForRenderer(
        <Comment commentId={1}/>,
        {preloadedState: createState(unlikedComment)});

    const tree = renderer
        .create(internalDom)
        .toJSON();

    expect(tree).toMatchSnapshot();
});


describe('like button clicked, when comment unliked', () => {
    test('should change to success color', async () => {
        // Arrange
        const likeButton = renderComment(unlikedComment);

        // Act
        fireEvent.click(likeButton);
        await new Promise(process.nextTick);

        // Assert
        expect(likeButton).toHaveStyle(`background-color: ${testTheme.successColor};`);
    });

    test('should increment likes number', async () => {
        // Arrange
        const likeButton = renderComment(unlikedComment);

        // Act
        fireEvent.click(likeButton);
        await new Promise(process.nextTick);

        // Assert
        const number = screen.getByText(unlikedComment.likes + 1);
        expect(number).toBeDefined();
    });
});

describe('like button clicked, when comment liked', () => {
    test('becomes primary color', async () => {
        // Arrange
        const likeButton = renderComment(likedComment);

        // Act
        fireEvent.click(likeButton);
        await new Promise(process.nextTick);

        // Assert
        expect(likeButton).toHaveStyle(`background-color: ${testTheme.primaryColor};`);
    });

    test('should decrement likes number', async () => {
        // Arrange
        const likeButton = renderComment(likedComment);

        // Act
        fireEvent.click(likeButton);
        await new Promise(process.nextTick);

        // Assert
        const number = screen.getByText(likedComment.likes - 1);
        expect(number).toBeDefined();
    });
});


test('clicks on the like button, when changing like status, should not trigger more changes', async () => {
    // Arrange
    const likeButton = renderComment(likedComment);

    // Act
    fireEvent.click(likeButton); // unlike
    fireEvent.click(likeButton); // like
    await new Promise(process.nextTick);

    // Assert
    // The status change was only triggered once making it unliked.
    expect(likeButton).toHaveStyle(`background-color: ${testTheme.primaryColor};`);
});