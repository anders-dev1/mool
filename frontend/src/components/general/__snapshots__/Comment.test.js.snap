// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="flex justify-start"
  >
    <img
      className="sc-bczRLJ ddxDEV mr1"
      src="http://placekitten.com/256/256"
    />
    <div
      className="flex-column pt1"
    >
      <p
        className="sub"
      >
        This is the comment content.
      </p>
      <p
        className="sub mt1"
        data-testid="commentContent"
      >
        Lorem ipsum
      </p>
    </div>
  </div>
  <div
    className="flex justify-end"
  >
    <div
      className="sc-hKMtZM gBUhPo mr1"
      color="#007BFF"
      data-testid="commentLikeBtn"
      href="#"
      onClick={[Function]}
    >
      <p
        className="sc-eCYdqJ grMfHa"
      >
        üëç
      </p>
      <div
        className="sc-jSMfEi fhnsns"
      >
        <p
          className="sc-gKXOVf knNWEu"
          data-testid="number"
        >
          42
        </p>
      </div>
    </div>
  </div>
</div>
`;
